packages/@lwc/engine-dom/src/apis/hydrate-component.ts

```diff
@@ -33,42 +66,43 @@ export function hydrateComponent(
         );
     }
 
+    if (getAssociatedVMIfPresent(element)) {
+        /* eslint-disable-next-line no-console */
+        console.warn(
+            `"hydrateComponent" expects an element that is not hydrated, instead received %o.`,
+            element
+        );
+        return;
+    }
+
     try {
         // Let the renderer know we are hydrating, so it does not replace the existing shadowRoot
         // and uses the same algo to create the stylesheets as in SSR.
         setIsHydrating(true);
 
-        createVM(element, Ctor, {
-            mode: 'open',
-            owner: null,
-            tagName: element.tagName.toLowerCase(),
-        });
-
-        for (const [key, value] of Object.entries(props)) {
-            (element as any)[key] = value;
-        }
+        createVMWithProps(element, Ctor, props);
 
         hydrateRootElement(element);
 
         // set it back since now we finished hydration.
         setIsHydrating(false);
     } catch (e) {
-        // Fallback: In case there's an error while hydrating, let's log the error, and replace the element with
-        //           the client generated DOM.
+        // Fallback: In case there's an error while hydrating, let's log the error, and replace the element content
+        //           with the client generated DOM.
 
         /* eslint-disable-next-line no-console */
         console.error('Recovering from error while hydrating: ', e);
 
-        setIsHydrating(false);
-        const newElem = createElement(element.tagName, {
-            is: Ctor,
-            mode: 'open',
-        });
+        // We want to preserve the element, so we need to reset the shadowRoot and light dom.
+        resetShadowRootAndLightDom(element, Ctor);
```

Comment:
What is the value of preserving the host element?

Reply:
For `hydrateComponent`, none. however, in the case of `customElements.define`, (imo) it feels weird that the `this` being constructed may be removed. I also like the idea of not handling anything other than the element.